src/TwabController.sol

```diff
@@ -423,48 +463,64 @@ contract TwabController {
     }
 
     // If we are transferring tokens from a delegated account to an undelegated account
+    address _fromDelegate = _delegateOf(_vault, _from);
+    address _toDelegate = _delegateOf(_vault, _to);
     if (_from != address(0)) {
-      address _fromDelegate = _delegateOf(_vault, _from);
       bool _isFromDelegate = _fromDelegate == _from;
 
       _decreaseBalances(_vault, _from, _amount, _isFromDelegate ? _amount : 0);
 
-      if (!_isFromDelegate) {
-        _decreaseBalances(
-          _vault,
-          _fromDelegate,
-          0,
-          _fromDelegate != SPONSORSHIP_ADDRESS ? _amount : 0
-        );
+      // If the user is not delegating to themself, increase the delegate's delegateBalance
+      // If the user is delegating to the sponsorship address, don't adjust the delegateBalance
+      if (!_isFromDelegate && _fromDelegate != SPONSORSHIP_ADDRESS) {
+        _decreaseBalances(_vault, _fromDelegate, 0, _amount);
       }
 
-      // burn
-      if (_to == address(0)) {
+      // Burn balance if we're transferring to address(0)
+      // Burn delegateBalance if we're transferring to address(0) and burning from an address that is not delegating to the sponsorship address
+      // Burn delegateBalance if we're transferring to an address delegating to the sponsorship address from an address that isn't delegating to the sponsorship address
+      if (
+        _to == address(0) ||
+        (_toDelegate == SPONSORSHIP_ADDRESS && _fromDelegate != SPONSORSHIP_ADDRESS)
+      ) {
+        // If the user is delegating to the sponsorship address, don't adjust the total supply delegateBalance
         _decreaseTotalSupplyBalances(
           _vault,
-          _amount,
-          _fromDelegate != SPONSORSHIP_ADDRESS ? _amount : 0
+          _to == address(0) ? _amount : 0,
+          (_to == address(0) && _fromDelegate != SPONSORSHIP_ADDRESS) ||
+            (_toDelegate == SPONSORSHIP_ADDRESS && _fromDelegate != SPONSORSHIP_ADDRESS)
+            ? _amount
+            : 0
         );
       }
     }
 
-    // If we are transferring tokens from an undelegated account to a delegated account
+    // If we are transferring tokens to an address other than address(0)
     if (_to != address(0)) {
-      address _toDelegate = _delegateOf(_vault, _to);
       bool _isToDelegate = _toDelegate == _to;
 
+      // If the user is delegating to themself, increase their delegateBalance
       _increaseBalances(_vault, _to, _amount, _isToDelegate ? _amount : 0);
 
-      if (!_isToDelegate) {
-        _increaseBalances(_vault, _toDelegate, 0, _toDelegate != SPONSORSHIP_ADDRESS ? _amount : 0);
+      // Otherwise, increase their delegates delegateBalance if it is not the sponsorship address
+      if (!_isToDelegate && _toDelegate != SPONSORSHIP_ADDRESS) {
+        _increaseBalances(_vault, _toDelegate, 0, _amount);
```

Comment:
`_transferDelegateBalance` is used when we delegate. 

If we called `_transferDelegateBalance` inside `_transferBalance` we would be duplicating all of the checks when we need to do them for the balance transfer. 
If the user is delegating to themself we only need to do one write to memory to update both values. If we always break up the updates to `balance` and `delegateBalance` we would need to do 2 writes if the user is delegating to themself.

Reply:
Yes, what I meant is that once you enter this condition, you will necessarily do two writes.
I feel like we duplicate the logic since we have the same kind of logic in `_transferDelegateBalance` but maybe we are ok with it.

Reply:
`_transferDelegateBalance` is basically subset of `_transferBalance`. For something like delegating, we can use the more optimized `_transferDelegateBalance` to remove some unnecessary checks.

Doing two writes is unavoidable if the user is delegating to someone other than themselves. But having `_transferBalance` manage updating both the `balance` & `delegateBalance` then we occasionally (likely most of the time since most users will be delegating to themselves) get to do a single write.